### PR TITLE
Wire up tidy dependency checks for cg_clif

### PR DESCRIPTION
**Pull request by bjorn3**
Also contains a fix and improvement to tidy.

Required for https://github.com/rust-lang/rust/pull/81746.

### CODE REVIEW

File: src/tools/tidy/src/deps.rs [hunk 1]
```diff
@@ -44,12 +44,29 @@ const EXCEPTIONS: &[(&str, &str)] = &[
     ("fortanix-sgx-abi", "MPL-2.0"), // libstd but only for `sgx` target
 ];
 
+const EXCEPTIONS_CRANELIFT: &[(&str, &str)] = &[
+    ("cranelift-bforest", "Apache-2.0 WITH LLVM-exception"),
```
Comment (bjorn3) [inline src/tools/tidy/src/deps.rs:48]: This license is used by LLVM too, so it shouldn't be a problem.

File: src/tools/tidy/src/deps.rs [hunk 2]
```diff
@@ -44,12 +44,29 @@ const EXCEPTIONS: &[(&str, &str)] = &[
...
+    ("cranelift-codegen-meta", "Apache-2.0 WITH LLVM-exception"),
+    ("cranelift-codegen-shared", "Apache-2.0 WITH LLVM-exception"),
+    ("cranelift-entity", "Apache-2.0 WITH LLVM-exception"),
+    ("cranelift-frontend", "Apache-2.0 WITH LLVM-exception"),
+    ("cranelift-jit", "Apache-2.0 WITH LLVM-exception"),
+    ("cranelift-module", "Apache-2.0 WITH LLVM-exception"),
+    ("cranelift-native", "Apache-2.0 WITH LLVM-exception"),
+    ("cranelift-object", "Apache-2.0 WITH LLVM-exception"),
+    ("libloading", "ISC"),
+    ("mach", "BSD-2-Clause"),
```
Comment (bjorn3) [inline src/tools/tidy/src/deps.rs:59]: Both of these licenses are equivalent to MIT I believe, but IANAL.

**Comment by joshtriplett** [comment]
Would it potentially make sense to deduplicate the `PERMITTED_CRANELIFT_DEPENDENCIES` list with the `PERMITTED_DEPENDENCIES` list, and just add comments (or actual enum-based data types perhaps) to the cranelift-only ones? I think that'd make the list much easier to maintain, and in particular, it would make sure people don't let the lists get out of sync.

I'm thinking of two cases here:

- Someone adds a dependency, and it'll definitely apply to all backends, but they don't realize they need to apply it to the cranelift list, so CI fails and they end up having to fix that later.
- Someone removes a dependency from one list, but it's actually no longer needed for any backend.

**Comment by bjorn3** [comment]
There are two lists as there are two workspaces. Changing a single `Cargo.toml` to add a dependency will only affect a single workspace and thus only one of the lists. The reason I use separate lists is because if I merge the lists, tidy will complain for both workspaces that there are exceptions that don't exist in the workspace that is currently being checked due to it only existing in the other workspace. This check also means that you can't forget to remove an exception fron either list once it is no longer necessary.

**Comment by joshtriplett** [comment]
@bjorn3 Would it be feasible to merge the lists into one list that tags each dependency as needed for one, the other, or both? That seems like it'd simplify maintenance, and avoid additional CI cycles ("oops, I forgot to fix the backend I'm not using").

**Comment by bjorn3** [comment]
Running tidy will always check both the main workspace and cg_clif's workspace. Merging both lists doesn't change anything about that. The only change will be that it makes the list a bit harder to read I think. There is barely any overlap in dependencies between rustc and cg_clif anyway.

**Comment by joshtriplett** [comment]
@bjorn3 Ah, I see; I didn't realize that both lists would always get checked. In that case, this isn't a semantic issue, only a question of what's easier to maintain.

Sorry for the noise, then.

**Comment by Mark-Simulacrum** [comment]
@bors r+

I think this PR is pretty much a strict improvement, though I want to be clear that this is not a confirmation of the license exceptions being OK - we'll want to evaluate those when we get closer to distributing cranelift, but for now this PR is still just an improvement of the status quo. Thanks!

**Pull request merged**